### DISCUSSION TITLE
[Testing Automation] Enhancing 3DS card payment for Adyen connector using Cypress

### DISCUSSION THREAD

**Discussion by neerajkumar-n** [juspay/hyperswitch#4316] [General]
**Context:**\
Ensuring the smooth operation of our payment processing system is crucial for the success of our application. One of the Connectors we rely on is Adyen payment gateway and we use this for handling transactions, focusing on both one-time payments & Subscription/repeat transactions.

Our current testing framework lacks comprehensive coverage for both one-time & Subscription payment scenarios with the Adyen connector. We currently use Newman for automation testing but have encountered challenges. Maintaining separate collections for different connectors has become burdensome, requiring manual verification of every flow for each connector with every change. Newman's inability to handle redirection flows, like 3DS flows, and its limitations in testing other payment methods such as wallets and bank transfers, further compound our testing efforts. Recognizing the inefficiency and limitations, we acknowledge the need to transition to Cypress.

By automating our testing processes using cypress, we aim to streamline workflows, improve efficiency, and address the shortcomings encountered with our current automation testing, ensuring a more robust and reliable testing strategy moving forward. We are committed to enhancing our testing strategy by incorporating dedicated test scenarios specifically targeting one-time payments within the Adyen connector. By doing so, we aim to ensure that all aspects of one-time payments, including transaction initiation, authorization, and confirmation, are thoroughly tested and validated for card payments.

In summary, by prioritizing the enhancement of our testing framework to include one-time payment scenarios, we reaffirm our commitment to quality, reliability, and continuous improvement in our payment processing ecosystem. This approach will enable us to identify and rectify potential issues before they impact our users' experience.

****

**How to get started**

- [Set up a Rust environment and run Hyperswitch locally](https://github.com/juspay/hyperswitch/blob/main/docs/try_local_system.md)

- Signup and creating a sandbox account on [Adyen](https://www.adyen.com/) to get the API Credentials

- Configure Connector within Hyperswitch dashboard, you can sign up for Hyperswitch [here](http://app.hyperswitch.io). You can learn more about configuring Connectors in our documentation [here](https://docs.hyperswitch.io/hyperswitch-cloud/account-setup).

- Once configured, you can simply test a payment via our dashboard, learn more about it [here](https://docs.hyperswitch.io/hyperswitch-open-source/testing/test-a-payment)

**How to Implement:**

- Read and understand the [general guidelines](https://github.com/juspay/hyperswitch/blob/main/docs/CONTRIBUTING.md) on contributing to Hyperswitch

- Import postman collection of hyperswitch from [here](https://www.postman.com/hyperswitch/workspace/hyperswitch-development/collection/25176162-630b5353-7002-44d1-8ba1-ead6c230f2e3) 

- You can refer to our Documentation [here](http://api-reference.hyperswitch.io)

****

**Expected Outcomes:**

Testing Procedure for Card Payment Authentication (authentication\_type: three\_ds)
- Initiate Card Payment with Three-DS Authentication:
  - Create a card payment and confirm it with the authentication\_type set as "three\_ds".

- Verification of Response Link:
  - Verify the link provided in the response.
  - Open the provided link to proceed with authentication.

- Complete Authentication and Redirect:
  - Follow the authentication process and complete it.
  - Redirect back to the return URL specified.

- Verification of Payment Status:
  - Verify the status of the payment after authentication completion.

- Failed Authentication Handling:
  - Simulate a failed authentication attempt.
  - Verify that the payment status reflects the failed authentication state.

Submit a pull request to the juspay/hyperswitch repository including the relevant screenshots of the test cases.\
Brownie Points: Providing clear and comprehensive documentation and code comments.\
Note: We welcome multiple PRs for this problem statement

****

**Additional Resources:** 

- If you face any issues/challenges while using Cypress, we recommend you to refer to [their documentation](https://docs.cypress.io/guides/overview/why-cypress)

- Here is a [test case](https://github.com/juspay/hyperswitch/tree/main/cypress-tests) that we have already worked upon for your reference.

- You can read more about 3DS [here](https://www.ebanx.com/en/resources/payments-explained/3d-secure/)

**Comment by Ovank** [comment]
Hey @neerajkumar-n , I am having issue in setting up Adyen test account as it is not accepting a gmail email, Do you have some suggestion to work around it.

**Reply by manav-chan** [reply]
Hi, I used my college email id for that task.

**Reply by neerajkumar-n** [reply]
Yes, college email id would work or orgs email id will also work.

**Reply by Tarun-Manjunath** [reply]
Hi, is there any deadline before which this task must be completed.

**Comment by trial-pyth** [comment]
While starting `docker-compose.yml` got the following error for `hyperswitch-server` service:
```
> Unable to deserialize application configuration: pm_filters.zsl.local_bank_transfer: data did not match any variant of untagged enum PaymentMethodFilterKey
> thread 'main' panicked at crates/router/src/bin/router.rs:14:10:
> Unable to construct application configuration: ConfigurationError(data did not match any variant of untagged enum PaymentMethodFilterKey)
```

While starting `docker-compose-development.yml` there was no indication of a server started on port `8080`

After following instructions given in the documentation file:
```
C:\Users\DELL>curl -I -X GET http://localhost:8080/health
curl: (52) Empty reply from server
```

Able to ping `hyperswitch-server` from another container, but not able to curl `http://hyperswitch-server:8080`

**Comment by Tarun-Manjunath** [comment]
Hi, after hosting my playground on my local machine. The playground doesn,t have any content at all.
I did every step that was told in docs.
cd hyperswitch-web
npm run start
npm run re:start
npm run build playground
npm run install playground

I even hosted my local server as well still I'm not getting my playground content
![65](https://github.com/juspay/hyperswitch/assets/91819064/0fa83d69-1dc6-4067-8395-76c41b95e756)
.

**Reply by PritishBudhiraja** [reply]
Hi @Tarun-Manjunath ,
Can you please check the .env file if you are doing local development.
```bash
STATIC_DIR=./dist
HYPERSWITCH_PUBLISHABLE_KEY=******
HYPERSWITCH_SECRET_KEY=*****
HYPERSWITCH_SERVER_URL="http:*******:8080/" - check if there any slash is there in the end if it's there please remove it
HYPERSWITCH_CLIENT_URL="http:*******9050/" - check if there any slash is there in the end if it's there please remove it
```

If any other issue is there let me know.

you can follow these commands if playground might not work for you run these commands on separate terminals

```bash
1. npm run re:start
2. npm run start
3. cd Hyperswitch-React-Demo-App && npm run start
```

Let me know if you still face any issues.

Thanks.

**Reply by Tarun-Manjunath** [reply]
Hi, @PritishBudhiraja well I changed my env variables but still couldn't get the content, what I believe is that there is a problem in compiling.
Well if you can see the content should displayed from dist from it is from / so how do I change that, because all the contents like cap, shirt is in dist.
![hp1](https://github.com/juspay/hyperswitch/assets/91819064/a8b83730-9240-4827-9dd8-75e74a27c157)
![hp2](https://github.com/juspay/hyperswitch/assets/91819064/a146ef51-20d9-40c8-b97d-45398104400f)

**Reply by PritishBudhiraja** [reply]
Hey can you please change the last 2 env values.
```bash
HYPERSWITCH_SERVER_URL="http://localhost:8080"
# At 9060 port our demo app is running which is giving the feel of SDK works.
# But at 9050 our SDK is running that is HyperLoader.js file. So that should be there in the client URL
HYPERSWITCH_CLIENT_URL="http://localhost:9050"
```

And if you are doing local backend setup please go to webpack.common.js  & webpack.dev.js- 
<img width="463" alt="Screenshot 2024-05-05 at 10 55 41 AM" src="https://github.com/juspay/hyperswitch/assets/77892330/9870e446-392e-4c0d-a707-c1de09d7f6ba">

<img width="463" alt="Screenshot 2024-05-05 at 10 59 23 AM" src="https://github.com/juspay/hyperswitch/assets/77892330/368d3b04-d0a8-465b-a31e-e3cb5f8ab93b">

Change the backend endpoint url to "http://localhost:8080" at both the places. let me know if you still face any issue.

**Reply by Tarun-Manjunath** [reply]
Hi @PritishBudhiraja I followed all the instructions but still couldn't get the content.
![hp7](https://github.com/juspay/hyperswitch/assets/91819064/a5dc758a-1c55-46f8-90a7-3cd99e34b551)

**Reply by Tarun-Manjunath** [reply]
![Uploading hp8.png…]()

**Reply by Tarun-Manjunath** [reply]
is there any variables I need to add in .env file

**Comment by neerajkumar-n** [comment]
Closing this discussion, this has been implemented.
